### PR TITLE
Add node navigation pages on /nodes/

### PR DESCRIPTION
**Pull request by catch-point**
Point your browser to http://localhost:8080/nodes/

### CODE REVIEW

**Comment by dbooth-boston** [comment]
I already have tomcat running on 8080, so I use port 3142 when I'm running noflo.  I tried:

```
noflo-server --gph graphs/montreal.json --port 3142
```

and the graph comes up in the GUI from http://localhost:3142/ but http://localhost:3142/nodes/ yields "Cannot GET /nodes/":

```
$ curl 'http://localhost:3142/nodes/'
Cannot GET /nodes/
$ curl -I 'http://localhost:3142/nodes/'
HTTP/1.1 404 Not Found
X-Content-Type-Options: nosniff
Content-Type: text/html; charset=utf-8
Content-Length: 20
Date: Wed, 11 May 2016 01:07:20 GMT
Connection: keep-alive
```

Should I try something else?

**Comment by glenna-m** [comment]
I am seeing the same thing David is seeing in the browser, but curl does work - I am getting back HTML content that matches my current graph.

**Comment by glenna-m** [comment]
Both curl and the browser are now working for me. 
screen shot 2016-05-10 at 10 15 45 pm
Nice work on this, James.

**Comment by catch-point** [comment]
What did you change it get it to work Glenna? Any tips that might help David?

File: src/navigator.js [hunk 1]
```diff
@@ -0,0 +1,154 @@
+// navigator.js
+
```
Comment (dbooth-boston) [inline src/navigator.js:2]: Suggest: Add a comment that documents the URI patterns (query strings) that this supports.  Or better yet: document them on the wiki, and then add a comment here that points to the wiki page.
https://github.com/rdf-pipeline/framework/wiki

**Comment by dbooth-boston** [comment]
Really nice!  Suggestions:
1. On non-RDF-Pipeline nodes, include something where the state/VNI would normally be displayed, saying why it is not displayed, such as "No state data (not an RDF Pipeline node)"
2. Regarding the table that lists the source and destination nodes, I think it will be easier to read if you break it into two tables, one above the other: one for input connections and one for output connections.
3. Same suggestion for the table of ports: break it into two tables (Inputs and Outputs).  It could all be physically part of the same table, but with an additional title row before the outputs.  (Input, Description) and (Output, Description).   Then the arrows won't be needed.
4. Is the 'label' property in the blue box redundant?  Is it always the same as the title?  If so it could be dropped.

**Comment by catch-point** [comment]
The 'label' property is redundant, but I think it makes the metadata box complete (since label is part of the node's metadata object)

File: handlebars/index.html.hbs [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <title>{{name}}</title>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
```
Comment (glenna-m) [inline handlebars/index.html.hbs:6]: Nice.

File: handlebars/index.html.hbs [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        <tbody>
+        {{#each nodes}}
+            <tr>
+                <td><a href="{{../path}}{{encodeURI id}}{{../search}}">{{id}}</a></td>
+                <td>{{component}}</td>
+        {{/each}}
+        </tbody>
+    </table>
+</div>
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js"></script>
```
Comment (glenna-m) [inline handlebars/index.html.hbs:30]: Would it make sense to bring jquery in as an npm package dependency?

File: handlebars/index.html.hbs [hunk 4]
```diff
@@ -0,0 +1,33 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <title>{{name}}</title>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+    <meta http-equiv="x-ua-compatible" content="ie=edge">
+    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-alpha.2/css/bootstrap.min.css" integrity="sha384-y3tfxAZXuh4HwSYylfB+J125MxIs6mR5FOHamPBG064zB+AFeWH94NdvaCBm8qnd" crossorigin="anonymous">
```
Comment (glenna-m) [inline handlebars/index.html.hbs:8]: Should we bring in the npm bootstrap package?

File: src/navigator.js [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+Handlebars.registerHelper('toJSON', function(object){
+    try {
+        return new Handlebars.SafeString(JSON.stringify(object, null, 2));
+    } catch (e) {
+        return new Handlebars.SafeString(e.toString());
+    }
+});
+
+Handlebars.registerHelper('encodeURI', function(uri){
+    return encodeURI(uri);
```
Comment (glenna-m) [inline src/navigator.js:18]: Good

File: src/noflo-server.js [hunk 6]
```diff
@@ -66,14 +67,16 @@ Promise.resolve(program.graph).then(function(filename){
...
     });
 }).then(function(graph){
-    return runtime(server, {
+    var rt = runtime(server, {
         defaultGraph: graph,
         baseDir: baseDir,
         captureOutput: false,
         catchExceptions: true,
         permissions: permissions,
         defaultPermissions: defaultPermissions
```
Comment (glenna-m) [inline src/noflo-server.js:76]: Right now, I don't think we have any special permissions, yes?  Do we have any requirements on that, @dbooth-boston ?

File: src/navigator.js [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+    if (vni) {
+        res.writeHead(200, {'Content-Type': 'application/json'});
+        res.end(JSON.stringify(vni.outputState, null, 2));
+    } else {
+        res.writeHead(404);
+        res.end("No output state available");
+    }
+}
+
+function view(network, facades, render, path, vnis, vnid, nodeId, qs, res) {
```
Comment (glenna-m) [inline src/navigator.js:175]: Should probably put some javadoc comments on this.  We all know what these parameters are, but those out side the project may not understand the facdes, or the vnis, vnid, nodeId...

**Comment by glenna-m** [comment]
LGTM pending the fairly minor suggestions.  Nice work.

File: src/navigator.js (hunk 7 above)
Comment (dbooth-boston) [inline src/navigator.js:175]: good idea

File: src/noflo-server.js (hunk 6 above)
Comment (dbooth-boston) [inline src/noflo-server.js:76]: None that we need to address at this point.

**Comment by dbooth-boston** [comment]
A couple of documentation items still need to be addressed, but aside from those, this looks good.

**Comment by dbooth-boston** [comment]
LGTM
